test_generators/operations/deposits.py

```diff
@@ -0,0 +1,168 @@
+
+    item = spec.hash(deposit_data.serialize())
+    index = len(deposit_data_leaves)
+    deposit_data_leaves.append(item)
+    tree = calc_merkle_tree_from_leaves(tuple(deposit_data_leaves))
+    proof = list(get_merkle_proof(tree, item_index=index))
+
+    deposit = spec.Deposit(
+        proof=list(proof),
+        index=index,
+        data=deposit_data,
+    )
+    assert spec.verify_merkle_branch(item, proof, spec.DEPOSIT_CONTRACT_TREE_DEPTH, index, get_merkle_root(tuple(deposit_data_leaves)))
+
+    return deposit
+
+
+def build_deposit_for_index(initial_validator_count: int, index: int) -> Tuple[spec.Deposit, spec.BeaconState]:
+    genesis_deposits = genesis.create_deposits(
+        keys.pubkeys[:initial_validator_count],
+        keys.withdrawal_creds[:initial_validator_count]
+    )
+    state = genesis.create_genesis_state(genesis_deposits)
+
+    deposit_data_leaves = [spec.hash(dep.data.serialize()) for dep in genesis_deposits]
+
+    deposit = build_deposit(
+        state,
+        deposit_data_leaves,
+        keys.pubkeys[index],
+        keys.withdrawal_creds[index],
+        keys.privkeys[index],
+        spec.MAX_DEPOSIT_AMOUNT,
+    )
+
+    state.latest_eth1_data.deposit_root = get_merkle_root(tuple(deposit_data_leaves))
+    state.latest_eth1_data.deposit_count = len(deposit_data_leaves)
+
+    return deposit, state
+
+
+@to_dict
+def valid_deposit():
+    new_dep, state = build_deposit_for_index(10, 10)
+    yield 'description', 'valid deposit to add new validator'
+    yield 'pre', encode(state, spec.BeaconState)
+    yield 'deposit', encode(new_dep, spec.Deposit)
+    spec.process_deposit(state, new_dep)
+    yield 'post', encode(state, spec.BeaconState)
+
+
+@to_dict
+def valid_topup():
+    new_dep, state = build_deposit_for_index(10, 3)
+    yield 'description', 'valid deposit to top-up existing validator'
+    yield 'pre', encode(state, spec.BeaconState)
+    yield 'deposit', encode(new_dep, spec.Deposit)
+    spec.process_deposit(state, new_dep)
+    yield 'post', encode(state, spec.BeaconState)
+
+
+@to_dict
+def invalid_deposit_index():
+    new_dep, state = build_deposit_for_index(10, 10)
+    # Mess up deposit index, 1 too small
+    state.deposit_index = 9
+
+    yield 'description', 'invalid deposit index'
+    yield 'pre', encode(state, spec.BeaconState)
+    yield 'deposit', encode(new_dep, spec.Deposit)
+    yield 'post', None
+
+@to_dict
+def invalid_deposit_proof():
+    new_dep, state = build_deposit_for_index(10, 10)
+    # Make deposit proof invalid (at bottom of proof)
+    new_dep.proof[-1] = spec.ZERO_HASH
+
+    yield 'description', 'invalid deposit proof'
+    yield 'pre', encode(state, spec.BeaconState)
+    yield 'deposit', encode(new_dep, spec.Deposit)
+    yield 'post', None
+
+
+@to_tuple
+def deposit_cases():
+    yield valid_deposit()
+    yield valid_topup()
+    yield invalid_deposit_index()
+    yield invalid_deposit_proof()
+
+
```

Comment:
Consider:

```python
def generate_deposit_suite(config: str, configs_path: str):
    presets = loader.load_presets(configs_path, config)
    ...
```

and then in `operations/main.py`:

```python
gen_runner.run_generator("operations", [
        generate_deposit_suite("minimal"),
        generate_deposit_suite("mainnet"),
])
```

Reply:
Your proposed solution could work, but requires currying (`generate_deposit_suite` returning a function that lets you run the test suite). Very nice and functional, but too much imho for something like this. I'd rather have the initial suites be explict about their intentions running with a given configuration. It's compatible, so if there ever arises a need for more configs, it should be easy enough to implement.